grammars/javascript.cson

```diff
@@ -281,240 +281,283 @@
+    'end': '''(?x)
+      (?<=})|
+      ((?!
+        \\s*{|
+        \\G\\(|
+        \\G[\\w$]+|
+        \\s*/\\*|//
+      )(?=\\s*\\S))
+    '''
     'patterns': [
       {
-        'include': '#function-params'
+        'begin': '\\G'
+        'end': '(?<=(=>))'
+        'name': 'meta.function.arrow.js'
+        'patterns': [
+          {
+            'include': '#arrow_function_innards'
+          }
+        ]
+      }
+      {
+        'include': '#function_body'
+      }
+      {
+        'include': '#comments'
       }
     ]
   }
   {
-    'begin': '(?:((\')([^\']*?)(\'))|((")([^"]*?)(")))\\s*(:)\\s*\\b(?:(async)(?:\\s+))?(function\\*?)\\s*(\\*?)\\s*([a-zA-Z_?$][\\w?$]*)?\\s*(\\()'
-    'beginCaptures':
-      '1':
-        'name': 'string.quoted.single.js'
-      '2':
-        'name': 'punctuation.definition.string.begin.js'
-      '3':
-        'name': 'entity.name.function.js'
-      '4':
-        'name': 'punctuation.definition.string.end.js'
-      '5':
-        'name': 'string.quoted.double.js'
-      '6':
-        'name': 'punctuation.definition.string.begin.js'
-      '7':
-        'name': 'entity.name.function.js'
-      '8':
-        'name': 'punctuation.definition.string.end.js'
-      '9':
-        'name': 'keyword.operator.assignment.js'
-      '10':
-        'name': 'storage.modifier.js'
-      '11':
-        'name': 'storage.type.function.js'
-      '12':
-        'name': 'storage.type.function.js'
-      '13':
-        'name': 'entity.name.function.js'
-      '14':
-        'name': 'punctuation.definition.parameters.begin.js'
-    'comment': 'Attempt to match "foo": function'
-    'end': '(\\))'
-    'endCaptures':
-      '1':
-        'name': 'punctuation.definition.parameters.end.js'
-    'name': 'meta.function.json.js'
+    # [.]foo = ... => ...
+    'begin': '''(?x)
+      (?=
+        (\\.)?[a-zA-Z_$][\\w$]*
+        \\s*(=)\\s*
+        ((\\(([^\\(\\)]*)?\\))|[\\w$]+)
+        \\s*=>
+      )
+    '''
+    'end': '''(?x)
+      (?<=})|
+      ((?!
+        \\s*{|
+        \\G(\\.)?[a-zA-Z_$][\\w$]*\\s*(=)\\s*\\(|
+        \\G(\\.)?[a-zA-Z_$][\\w$]*\\s*(=)\\s*[\\w$]+|
+        \\s*/\\*|//
+      )(?=\\s*\\S))
+    '''
     'patterns': [
       {
-        'include': '#function-params'
+        'begin': '\\G'
+        'end': '(?<=(=>))'
+        'name': 'meta.function.arrow.js'
+        'patterns': [
+          {
+            'match': '\\G(\\.)?([a-zA-Z_$][\\w$]*)\\s*(=)'
+            'captures':
+              '1':
+                'name': 'meta.delimiter.method.period.js'
+              '2':
+                'name': 'entity.name.function.js'
+              '3':
+                'name': 'keyword.operator.assignment.js'
+          }
+          {
+            'include': '#arrow_function_innards'
+          }
+        ]
+      }
+      {
+        'include': '#function_body'
+      }
+      {
+        'include': '#comments'
       }
     ]
   }
   {
-    'begin': '\\b(constructor)\\s*(\\()'
+    'match': '(=>)'
```

Comment:
Dang.  We really need to try to make all these function matches newline-aware.  I'll create a new issue for that.